bindings/rust/src/lib.rs

```diff
@@ -177,58 +116,32 @@ impl blst_fp12 {
             panic!("inputs' lengths mismatch");
         }
 
-        let pool = mt::da_pool();
-
-        let mut n_workers = pool.max_count();
-        if n_workers == 1 {
-            let qs: [*const _; 2] = [&q[0], ptr::null()];
-            let ps: [*const _; 2] = [&p[0], ptr::null()];
-            let mut out = MaybeUninit::<blst_fp12>::uninit();
-            unsafe {
-                blst_miller_loop_n(out.as_mut_ptr(), &qs[0], &ps[0], n_elems);
-                return out.assume_init();
-            }
-        }
-
-        let (tx, rx) = channel();
-        let counter = Arc::new(AtomicUsize::new(0));
-
-        let stride = core::cmp::min((n_elems + n_workers - 1) / n_workers, 16);
-        n_workers = core::cmp::min((n_elems + stride - 1) / stride, n_workers);
-        for _ in 0..n_workers {
-            let tx = tx.clone();
-            let counter = counter.clone();
+        let num_threads = rayon::current_num_threads();
+        let stride =
+            core::cmp::min((n_elems + num_threads - 1) / num_threads, 16);
 
-            pool.joined_execute(move || {
-                let mut acc = blst_fp12::default();
+        (0..n_elems)
+            .into_par_iter()
+            .step_by(stride)
+            .zip(q.par_chunks(stride).zip(p.par_chunks(stride)))
+            .map(|(start_offset, (q, p))| {
                 let mut tmp = MaybeUninit::<blst_fp12>::uninit();
-                let mut qs: [*const _; 2] = [ptr::null(), ptr::null()];
-                let mut ps: [*const _; 2] = [ptr::null(), ptr::null()];
+                let qs = [q.as_ptr(), ptr::null()];
+                let ps = [p.as_ptr(), ptr::null()];
 
-                loop {
-                    let work = counter.fetch_add(stride, Ordering::Relaxed);
-                    if work >= n_elems {
-                        break;
-                    }
-                    let n = core::cmp::min(n_elems - work, stride);
-                    qs[0] = &q[work];
-                    ps[0] = &p[work];
-                    unsafe {
-                        blst_miller_loop_n(tmp.as_mut_ptr(), &qs[0], &ps[0], n);
-                        acc *= tmp.assume_init();
-                    }
+                let n = core::cmp::min(n_elems - start_offset, stride);
+                unsafe {
+                    blst_miller_loop_n(
+                        tmp.as_mut_ptr(),
+                        qs.as_ptr(),
+                        ps.as_ptr(),
+                        n,
+                    );
+                    tmp.assume_init()
                 }
-
-                tx.send(acc).expect("disaster");
-            });
-        }
-
-        let mut acc = rx.recv().unwrap();
-        for _ in 1..n_workers {
-            acc *= rx.recv().unwrap();
-        }
-
-        acc
+            })
+            .reduce_with(|a, b| a * b)
```

Comment:
The reduction here is tricky to get fast because the miller_loop works in burst of work:
- Up until 8, the only work is copying, this is at most 1ns.
- At 8 or when sending to another thread, you trigger a Miller Loop that requires ~200us. A ratio of 200000x compared to copying.

The original algorithm has strictly lower bounds on actual work done.
It also has better load-balancing because if one thread is very busy, another can pick up the work that was not scheduled.

The main issue in the original is the cost of many threads calling `fetch_add` but it's still dwarfed by a single fp12_mul

Reply:
This is an earlier piece of this PR, I can rewrite it with `spawn_broadcast` like in other places to be closer to original, thanks!